File: Library/DSA Reference/Data Structures/2) Strings/strings.cpp
#include <bits/stdc++.h>
using namespace std;

/*
    ============================================================
    STRINGS CHEAT SHEET — C++20
    ============================================================

    CORE / MUST KNOW
    ----------------
    Main DSA string type: string

    Access character       O(1)
    Update character       O(1)
    length / size          O(1)
    empty                  O(1)
    Search                 O(n) typical / algorithm-dependent
    Concatenate            O(n + m)
    Substring              O(k)
    Insert / erase         O(n)
    Compare                O(min(n,m))
    Push back              O(1) amortized
    Pop back               O(1)
    Reverse                O(n)
    Sort characters        O(n log n)

    EXTRAS / REFERENCE
    ------------------
    rfind, replace, insert overloads, transform, count,
    min/max element, numeric conversion, character classification,
    iterator algorithms, manual substring search, etc.
*/

int main() {

    // ============================================================
    // CORE — MUST KNOW
    // ============================================================

    // ------------------------------------------------------------
    // 1. DECLARATION / INITIALISATION
    // ------------------------------------------------------------

    string a;
    a = "hello";

    string b = "hello world";
    string c("hello world");

    string d(5, 'a');                 // "aaaaa"

    string copied = b;                // copy

    // C-style character array / string literal:
    const char raw[] = "hello";

    // ------------------------------------------------------------
    // 2. INPUT
    // ------------------------------------------------------------

    /*
        cin >> inp
            Reads one whitespace-delimited token.

        getline(cin, inp)
            Reads an entire line including spaces.

        If getline() follows formatted extraction such as
        cin >> x, consume the leftover newline first.
    */

    string inp;

    // cin >> inp;

    // Correct general pattern after cin >> something:
    cin.ignore(numeric_limits<streamsize>::max(), '\n');
    // getline(cin, inp);

    // Read one character, including whitespace:
    // char ch = cin.get();

    // ------------------------------------------------------------
    // 3. SIZE / LENGTH / EMPTY
    // ------------------------------------------------------------

    int n = (int)b.size();             // O(1)
    int n2 = (int)b.length();          // O(1)
    bool isEmpty = b.empty();          // O(1)

    cout << "Size: " << n << '\n';
    cout << "Length: " << n2 << '\n';
    cout << "Empty: " << boolalpha << isEmpty << '\n';

    // ------------------------------------------------------------
    // 4. CHARACTER ACCESS
    // ------------------------------------------------------------

    char ch1 = b[0];                   // O(1)
    char ch2 = b.at(0);                // O(1), bounds checked

    if (!b.empty()) {
        char first = b.front();        // O(1)
        char last = b.back();          // O(1)

        cout << ch1 << ' ' << ch2 << '\n';
        cout << first << ' ' << last << '\n';
    }

    // C++ has no Python-style negative indexing.
    // b[-1] does NOT mean the last character.

    // ------------------------------------------------------------
    // 5. UPDATE CHARACTER
    // ------------------------------------------------------------

    if (!b.empty())
        b[0] = 'H';                    // O(1)

    // ------------------------------------------------------------
    // 6. CONCATENATION / APPEND
    // ------------------------------------------------------------

    string s1 = "abc";
    string s2 = "def";

    s1 = s1 + s2;                      // O(n + m)

    s1 += "ghi";                        // append string
    s1.append("jkl");                   // append string
    s1.push_back('!');                  // append one char
    s1.append(3, '?');                  // append repeated chars

    // ------------------------------------------------------------
    // 7. CLEAR
    // ------------------------------------------------------------

    s1.clear();
    // s1 = "";

    // ------------------------------------------------------------
    // 8. COMPARISON
    // ------------------------------------------------------------

    string x = "Rishabh";
    string y = "Ramdhave";

    bool equal = (x == y);
    bool notEqual = (x != y);
    bool less = (x < y);
    bool greater = (x > y);

    int cmp = x.compare(y);

    /*
        Comparison is lexicographical.

        compare():
            < 0  -> x < y
              0  -> x == y
            > 0  -> x > y
    */

    cout << "Equal: " << equal << '\n';
    cout << "Not equal: " << notEqual << '\n';
    cout << "compare: " << cmp << '\n';

    // TC: O(min(n,m))

    // ------------------------------------------------------------
    // 9. SUBSTRING
    // ------------------------------------------------------------

    string full = "rishabh shital ramdhave";

    string part = full.substr(5, 6);
    string rest = full.substr(5);

    // TC: O(k)
    // SC: O(k) for resulting string

    // ------------------------------------------------------------
    // 10. FIND
    // ------------------------------------------------------------

    string text = "hello world";

    size_t pos = text.find("world");

    if (pos != string::npos)
        cout << "'world' found at index " << pos << '\n';

    size_t charPos = text.find('o');

    // Search from an index:
    size_t from = text.find('o', 5);

    // Last occurrence:
    size_t last = text.rfind('o');

    /*
        If not found:
            string::npos

        find() supports both characters and substrings.
    */

    // ------------------------------------------------------------
    // 11. INSERT
    // ------------------------------------------------------------

    string tst2 = "rishabh ramdhave";

    tst2.insert(8, "shital ");
    // "rishabh shital ramdhave"

    // Insert repeated characters:
    tst2.insert(0, 3, '*');

    // TC: generally O(n + inserted length)

    // ------------------------------------------------------------
    // 12. ERASE
    // ------------------------------------------------------------

    string tst1 = "rishabh shital ramdhave";

    tst1.erase(8, 7);
    // "rishabh ramdhave"

    // Erase from position to end:
    tst1.erase(8);

    // Erase one character:
    if (!tst1.empty())
        tst1.erase(0, 1);

    // TC: O(n) in general because remaining characters may shift

    // ------------------------------------------------------------
    // 13. REPLACE
    // ------------------------------------------------------------

    string tst3 = "rishabh ramdhave";

    tst3.replace(0, 7, "samruddhi");

    // Replace one character with another:
    string rep = "apples and bananas";

    std::replace(
        rep.begin(),
        rep.end(),
        'a',
        'o'
    );

    cout << "Character-replaced: " << rep << '\n';
    // "opples ond bononos"

    /*
        IMPORTANT:
        std::replace() replaces individual elements.

        string::replace() replaces a substring/range.
    */

    // ------------------------------------------------------------
    // 14. PUSH / POP
    // ------------------------------------------------------------

    tst3.push_back('a');
    tst3.pop_back();

    // push_back: O(1) amortized
    // pop_back : O(1)

    // ------------------------------------------------------------
    // 15. TRAVERSAL
    // ------------------------------------------------------------

    cout << "Index traversal: ";
    for (int i = 0; i < (int)text.size(); ++i)
        cout << text[i] << ' ';
    cout << '\n';

    cout << "Range traversal: ";
    for (char ch : text)
        cout << ch << ' ';
    cout << '\n';

    cout << "Iterator traversal: ";
    for (auto it = text.begin(); it != text.end(); ++it)
        cout << *it << ' ';
    cout << '\n';

    cout << "Reverse traversal: ";
    for (auto it = text.rbegin(); it != text.rend(); ++it)
        cout << *it << ' ';
    cout << '\n';

    // All traversals: TC O(n), SC O(1) auxiliary

    // ------------------------------------------------------------
    // 16. REVERSE
    // ------------------------------------------------------------

    string rev = "abcdef";

    reverse(rev.begin(), rev.end());

    // TC: O(n)
    // SC: O(1)

    // ------------------------------------------------------------
    // 17. SORT
    // ------------------------------------------------------------

    string sorted = "dcba";

    sort(sorted.begin(), sorted.end());
    // "abcd"

    sort(sorted.rbegin(), sorted.rend());
    // "dcba"

    // TC: O(n log n)

    // ------------------------------------------------------------
    // 18. COUNT
    // ------------------------------------------------------------

    int countA = count(
        rep.begin(),
        rep.end(),
        'o'
    );

    // TC: O(n)

    // ============================================================
    // EXTRAS — REFERENCE
    // ============================================================

    // ------------------------------------------------------------
    // 19. CASE CONVERSION
    // ------------------------------------------------------------

    string str = "hello world";

    transform(
        str.begin(),
        str.end(),
        str.begin(),
        [](unsigned char ch) {
            return (char)toupper(ch);
        }
    );

    cout << "Uppercase: " << str << '\n';

    transform(
        str.begin(),
        str.end(),
        str.begin(),
        [](unsigned char ch) {
            return (char)tolower(ch);
        }
    );

    cout << "Lowercase: " << str << '\n';

    // TC: O(n), SC: O(1) auxiliary

    // ------------------------------------------------------------
    // 20. MIN / MAX CHARACTER
    // ------------------------------------------------------------

    string letters = "dcab";

    char mn = *min_element(
        letters.begin(),
        letters.end()
    );

    char mx = *max_element(
        letters.begin(),
        letters.end()
    );

    cout << "Min char: " << mn << '\n';
    cout << "Max char: " << mx << '\n';

    // TC: O(n), SC: O(1)

    // ------------------------------------------------------------
    // 21. NUMERIC STRING CONVERSION
    // ------------------------------------------------------------

    string numString = "12345";

    int integerValue = stoi(numString);
    long long longValue = stoll(numString);
    double decimalValue = stod("12.34");

    string backToString = to_string(12345);

    cout << integerValue << ' '
         << longValue << ' '
         << decimalValue << ' '
         << backToString << '\n';

    // ------------------------------------------------------------
    // 22. CHARACTER CLASSIFICATION
    // ------------------------------------------------------------

    char q = '7';

    bool alpha = isalpha((unsigned char)q);
    bool digit = isdigit((unsigned char)q);
    bool lower = islower((unsigned char)q);
    bool upper = isupper((unsigned char)q);
    bool alnum = isalnum((unsigned char)q);
    bool whitespace = isspace((unsigned char)q);

    cout << alpha << ' '
         << digit << ' '
         << lower << ' '
         << upper << ' '
         << alnum << ' '
         << whitespace << '\n';

    // ------------------------------------------------------------
    // 23. COPY / ASSIGN / SWAP
    // ------------------------------------------------------------

    string original = "hello";

    string copied2 = original;            // O(n)

    string assigned;
    assigned = original;                  // O(n)

    string other = "world";
    original.swap(other);                 // constant-time operation

    // ------------------------------------------------------------
    // 24. DATA / ITERATORS
    // ------------------------------------------------------------

    string dataString = "hello";

    auto beginIt = dataString.begin();
    auto endIt = dataString.end();

    const char* ptr = dataString.data();

    cout << "data(): " << ptr << '\n';

    // ------------------------------------------------------------
    // 25. ACCUMULATE
    // ------------------------------------------------------------

    string digits = "12345";

    int digitSum = accumulate(
        digits.begin(),
        digits.end(),
        0,
        [](int sum, char ch) {
            return sum + (ch - '0');
        }
    );

    cout << "Digit sum: " << digitSum << '\n';

    // Generic accumulate can also be adapted for other reductions.

    // ------------------------------------------------------------
    // 26. PARTIAL SUM
    // ------------------------------------------------------------

    vector<int> numbers = {1, 2, 3, 4};
    vector<int> prefix(numbers.size());

    partial_sum(
        numbers.begin(),
        numbers.end(),
        prefix.begin()
    );

    // prefix = {1,3,6,10}
    // TC: O(n), SC: O(n) output

    // ------------------------------------------------------------
    // 27. MANUAL LINEAR CHARACTER SEARCH
    // ------------------------------------------------------------

    auto findIndex = [](const string& s, char target) -> int {
        for (int i = 0; i < (int)s.size(); ++i) {
            if (s[i] == target)
                return i;
        }
        return -1;
    };

    cout << "Manual char search: "
         << findIndex("abcdef", 'd') << '\n';

    // TC: O(n)
    // SC: O(1)

    // ------------------------------------------------------------
    // 28. MANUAL PALINDROME
    // ------------------------------------------------------------

    auto isPalindrome = [](const string& s) -> bool {
        int l = 0;
        int r = (int)s.size() - 1;

        while (l < r) {
            if (s[l] != s[r])
                return false;

            ++l;
            --r;
        }

        return true;
    };

    cout << "Palindrome: "
         << isPalindrome("madam") << '\n';

    // TC: O(n)
    // SC: O(1)

    // ------------------------------------------------------------
    // 29. MANUAL REVERSE
    // ------------------------------------------------------------

    auto manualReverse = [](string& s) {
        int l = 0;
        int r = (int)s.size() - 1;

        while (l < r)
            swap(s[l++], s[r--]);
    };

    string manual = "abcdef";
    manualReverse(manual);

    // TC: O(n)
    // SC: O(1)

    // ------------------------------------------------------------
    // 30. NAIVE SUBSTRING SEARCH
    // ------------------------------------------------------------

    auto naiveFind = [](const string& s,
                        const string& pattern) -> int {

        if (pattern.empty())
            return 0;

        if (pattern.size() > s.size())
            return -1;

        for (int i = 0;
             i + (int)pattern.size() <= (int)s.size();
             ++i) {

            int j = 0;

            while (j < (int)pattern.size() &&
                   s[i + j] == pattern[j]) {
                ++j;
            }

            if (j == (int)pattern.size())
                return i;
        }

        return -1;
    };

    cout << "Naive substring search: "
         << naiveFind("hello world", "world") << '\n';

    // Worst-case TC: O(n*m)
    // SC: O(1)

    // ------------------------------------------------------------
    // 31. IMPORTANT LIBRARY NOTES
    // ------------------------------------------------------------

    /*
        string::find()
            Searches for a character or substring.

        std::find()
            Generic iterator algorithm; searches for an element,
            not a substring pattern.

        std::replace()
            Replaces elements/characters.

        string::replace()
            Replaces a range with a string/range.

        random_shuffle()
            Removed from modern C++.
            Use shuffle() instead.

        string::npos
            Sentinel meaning "not found".

        C++ string has mutable contiguous storage in standard
        implementations / guarantees relevant to data access,
        so data() provides access to its character buffer.
    */

    return 0;
}

File: Library/DSA Reference/Data Structures/2) Strings/strings.py
"""
    ============================================================
    STRINGS CHEAT SHEET — Python 3.x
    ============================================================

    CORE / MUST KNOW
    ----------------
    Main DSA string type: str

    Access character       O(1)
    len()                  O(1)
    Search                 O(n) typical
    Concatenate            O(n + m)
    Slice                  O(k)
    Compare                O(min(n,m))
    upper/lower            O(n)
    replace                O(n + output)
    count                  O(n)
    reverse copy           O(n)

    IMPORTANT DIFFERENCE:
        Python strings are IMMUTABLE.
        String "modification" creates a new string.

    EXTRAS / REFERENCE
    ------------------
    split, join, strip, classification, formatting, sorting,
    aliasing/identity, manual algorithms, etc.
"""

import re


# ================================================================
# CORE — MUST KNOW
# ================================================================

# ------------------------------------------------
# 1. INITIALISATION
# ------------------------------------------------

name = "Rishabh"

a = "hello"
b = 'hello world'

# Multiline:
multi = """
hello
world
"""

# Raw string:
raw = r"C:\Users\name\file.txt"


# ------------------------------------------------
# 2. INPUT
# ------------------------------------------------

# input() returns a string.
#
# s = input()
# s = input().strip()
# parts = input().split()

sample = "10 20 30"
parts = sample.split()

numbers = list(map(int, parts))


# ------------------------------------------------
# 3. SIZE / EMPTY
# ------------------------------------------------

s = "hello"

n = len(s)                    # O(1)
is_empty = not s              # O(1)

print("Length:", n)
print("Empty:", is_empty)


# ------------------------------------------------
# 4. CHARACTER ACCESS
# ------------------------------------------------

first = s[0]                  # O(1)
last = s[-1]                  # O(1)
second_last = s[-2]           # O(1)

# Negative indexing is supported.


# ------------------------------------------------
# 5. IMMUTABILITY / CHARACTER UPDATE
# ------------------------------------------------

# s[0] = "H"                  # TypeError

# Correct:
s = "H" + s[1:]

# TC: O(n)


# ------------------------------------------------
# 6. CONCATENATION / REPETITION
# ------------------------------------------------

s1 = "abc"
s2 = "def"

combined = s1 + s2            # O(n + m)

s1 += s2                       # Creates a new string conceptually

repeated = "abc" * 3           # O(n*k)


# ------------------------------------------------
# 7. CLEAR
# ------------------------------------------------

s = "hello"
s = ""

# Strings are immutable.
# Rebinding does not mutate an existing string object.


# ------------------------------------------------
# 8. COMPARISON
# ------------------------------------------------

a = "Rishabh"
b = "Ramdhave"

print(a == b)
print(a != b)

print(a < b)
print(a <= b)
print(a > b)
print(a >= b)

# Lexicographical comparison:
# TC: O(min(n,m))

# IMPORTANT:
# a is b checks object identity, NOT string value equality.


# ------------------------------------------------
# 9. SUBSTRING / SLICING
# ------------------------------------------------

s = "rishabh shital ramdhave"

part = s[5:11]
rest = s[5:]
prefix = s[:7]
suffix = s[-7:]

every_second = s[::2]
reversed_s = s[::-1]

# Slice creates a new string.
# TC: O(k), SC: O(k)


# ------------------------------------------------
# 10. SEARCH
# ------------------------------------------------

text = "hello world"

pos = text.find("world")        # -1 if absent
char_pos = text.find("o")
last_pos = text.rfind("o")

# index() raises ValueError if absent:
try:
    idx = text.index("world")
except ValueError:
    idx = -1

# Membership:
present = "world" in text


# ------------------------------------------------
# 11. REPLACEMENT
# ------------------------------------------------

s = "apple apple banana"

result = s.replace("apple", "orange")

# Limit replacements:
limited = s.replace("apple", "orange", 1)

# Original string remains unchanged.


# ------------------------------------------------
# 12. CASE
# ------------------------------------------------

s = "hello WORLD"

upper = s.upper()
lower = s.lower()
capitalized = s.capitalize()
title = s.title()

# Each returns a new string.


# ------------------------------------------------
# 13. TRAVERSAL
# ------------------------------------------------

s = "hello"

for ch in s:
    print(ch)

for i in range(len(s)):
    print(i, s[i])

for i, ch in enumerate(s):
    print(i, ch)

for ch in reversed(s):
    print(ch)

# TC: O(n)


# ------------------------------------------------
# 14. COUNT
# ------------------------------------------------

count_world = "hello world world".count("world")
count_a = "banana".count("a")

# TC: O(n) typical


# ================================================================
# EXTRAS — REFERENCE
# ================================================================

# ------------------------------------------------
# 15. SPLIT / REGEX SPLIT
# ------------------------------------------------

text = "apple banana grape"

words = text.split()

csv = "apple,banana,grape"
items = csv.split(",")

# Regex split:
text = "apple;banana,grape orange"
result = re.split(r"[;, ]+", text)

# TC: O(n) typical
# SC: proportional to produced output


# ------------------------------------------------
# 16. STRIP
# ------------------------------------------------

s = "   hello world   "

clean = s.strip()
left = s.lstrip()
right = s.rstrip()

custom = "---hello---".strip("-")

# ------------------------------------------------
# 17. JOIN
# ------------------------------------------------

words = ["hello", "world"]

joined = " ".join(words)
csv = ",".join(words)

# TC: O(total input size + result size)
# SC: O(result size)

# IMPORTANT:
# join() is the preferred way to combine many strings.


# ------------------------------------------------
# 18. CHARACTER / STRING CLASSIFICATION
# ------------------------------------------------

s = "Hello"

is_alpha = s.isalpha()
is_digit = s.isdigit()
is_lower = s.islower()
is_upper = s.isupper()
is_alnum = s.isalnum()
is_space = s.isspace()

# Other useful methods:
is_ascii = s.isascii()
is_title = s.istitle()
is_identifier = s.isidentifier()


# ------------------------------------------------
# 19. FORMATTING
# ------------------------------------------------

name = "World"

formatted1 = "Hello {}!".format(name)

formatted2 = f"Hello {name}!"

# Modern Python usually prefers f-strings.


# ------------------------------------------------
# 20. SORTING CHARACTERS
# ------------------------------------------------

s = "dcba"

sorted_chars = sorted(s)              # ['a','b','c','d']
sorted_string = "".join(sorted(s))    # "abcd"

reverse_sorted = "".join(
    sorted(s, reverse=True)
)

# TC: O(n log n)
# SC: O(n)

# Strings have no in-place .sort() method.


# ------------------------------------------------
# 21. MIN / MAX
# ------------------------------------------------

s = "dcab"

mn = min(s)
mx = max(s)

# TC: O(n)


# ------------------------------------------------
# 22. CONVERSION
# ------------------------------------------------

integer_value = int("12345")
float_value = float("12.34")

string_value = str(12345)


# ------------------------------------------------
# 23. MANUAL CHARACTER SEARCH
# ------------------------------------------------

def find_index(s, target):
    for i, ch in enumerate(s):
        if ch == target:
            return i
    return -1


print("Manual char search:",
      find_index("abcdef", "d"))

# TC: O(n)
# SC: O(1)


# ------------------------------------------------
# 24. MANUAL PALINDROME
# ------------------------------------------------

def is_palindrome(s):
    left, right = 0, len(s) - 1

    while left < right:
        if s[left] != s[right]:
            return False

        left += 1
        right -= 1

    return True


print("Palindrome:",
      is_palindrome("madam"))

# TC: O(n)
# SC: O(1)


# ------------------------------------------------
# 25. MANUAL REVERSE
# ------------------------------------------------

def reverse_string(s):
    chars = list(s)

    left, right = 0, len(chars) - 1

    while left < right:
        chars[left], chars[right] = (
            chars[right], chars[left]
        )
        left += 1
        right -= 1

    return "".join(chars)


print("Manual reverse:",
      reverse_string("abcdef"))

# TC: O(n)
# SC: O(n), because Python str is immutable


# ------------------------------------------------
# 26. NAIVE SUBSTRING SEARCH
# ------------------------------------------------

def naive_find(text, pattern):
    if pattern == "":
        return 0

    if len(pattern) > len(text):
        return -1

    for i in range(
        len(text) - len(pattern) + 1
    ):
        j = 0

        while (
            j < len(pattern)
            and text[i + j] == pattern[j]
        ):
            j += 1

        if j == len(pattern):
            return i

    return -1


print(
    "Naive substring search:",
    naive_find("hello world", "world")
)

# Worst-case TC: O(n*m)
# SC: O(1)


# ------------------------------------------------
# 27. ALIASING / IDENTITY
# ------------------------------------------------

a = "hello"
b = a

# Strings are immutable, so mutating b cannot change a.

# Value comparison:
same_value = (a == b)

# Object identity:
same_object = (a is b)

# Do not use `is` for normal string-value comparison.


# ------------------------------------------------
# 28. COMMON DSA STRING BUILDING PATTERN
# ------------------------------------------------

parts = []

for i in range(5):
    parts.append(str(i))

result = "".join(parts)

# Better than repeated concatenation for building many pieces.


# ------------------------------------------------
# 29. COMMON PYTHON STRING METHODS
# ------------------------------------------------

"""
Length:
    len(s)

Access:
    s[i]
    s[-1]

Search:
    s.find(x)
    s.index(x)
    s.rfind(x)
    x in s

Case:
    s.upper()
    s.lower()
    s.capitalize()
    s.title()

Whitespace:
    s.strip()
    s.lstrip()
    s.rstrip()

Replacement:
    s.replace(old, new)
    s.replace(old, new, count)

Counting:
    s.count(x)

Split:
    s.split()
    s.split(sep)

Join:
    sep.join(iterable)

Checks:
    s.isalpha()
    s.isdigit()
    s.isalnum()
    s.islower()
    s.isupper()
    s.isspace()

Formatting:
    s.format(...)
    f"..."

Sorting:
    sorted(s)

Reverse:
    s[::-1]
"""


# ------------------------------------------------
# 30. IMPORTANT PYTHON NOTES
# ------------------------------------------------

"""
1. str is immutable.
2. s[i] is O(1).
3. Negative indexing is supported.
4. Slicing creates a new string.
5. s.find(x) returns -1 if absent.
6. s.index(x) raises ValueError if absent.
7. x in s checks membership.
8. replace() handles substring replacement.
9. join() is preferred for joining many strings.
10. Strings have no .sort() method.
11. sorted(s) returns a list.
12. `is` is identity, `==` is value equality.
13. Avoid repeated large-string concatenation in a loop;
    collect pieces and use join().
"""
